be1-go/hub/organizer.go

```diff
@@ -53,7 +53,16 @@ func (o *organizerHub) Recv(msg IncomingMessage) {
 func (o *organizerHub) handleIncomingMessage(incomingMessage *IncomingMessage) {
 	log.Printf("organizerHub::handleIncomingMessage: %s", incomingMessage.Message)
 
-	client := incomingMessage.Client
+	socket := incomingMessage.Socket
+	client := ClientSocket{
+		Socket{
+			socketType: clientSocket,
+			hub:        socket.hub,
+			conn:       socket.conn,
+			send:       socket.send,
+			Wait:       sync.WaitGroup{},
+		},
+	}
```

Comment:
Shouldn't handling of incoming messages depend on the socket type `incomingmessage.Socket.socketType` in the message? What I have in mind is a switch statement based on the socket type. Here you're straight away assuming the message came on a client socket, why is that?

Reply:
Another question: why not call client := NewClientSocket(...)` (defined in `socket_types.go` in this same PR) instead of creating it by hand?

Reply:
The handling of messages should indeed depend on the socket type and I will definitely add a switch! I did not yet as I just wanted the refactored version to work, and for now the only sockets used are clients but I will add a switch anyway, as it will be useful for future code.
I do not call NewClientSocket here as NewClientSocket creates a new send channel and I thought that the client variable should keep the same one. I'm not quite sure though whether this makes a difference, do you know?
Also instead of using type composition to create different types of sockets, I am wondering whether I should just keep the socket type and use the "socketType" attribute in order to tell them apart as there is no difference in the way they function. What do you think?

Reply:
Yes, I think adding a switch where the not-yet-implemented branches throw an error "Not implemented" or similar would be very useful in preventing bugs. Thanks!

Reply:
You're right, here we don't want a new channel, which is what NewClientSocket does. So perhaps it's fine to leave it like this for now.

Reply:
_Also instead of using type composition to create different types of sockets, I am wondering whether I should just keep the socket type and use the "socketType" attribute in order to tell them apart as there is no difference in the way they function. What do you think?_ Good point. I'm not sure though what's best from a coding practice point of view, so I'd like to ask for @gnarula's opinion. Gaurav, could I ask what you think about it? Thanks!

Reply:
I'd suggest something like:

```go
type Socket interface {
  ReadPump()
  WritePump()
  Type() SocketType
  // .. other methods
}

// baseSocket implements Socket interface
type baseSocket struct {
  // fields go in here
}

// ClientSocket represents a connection with a client.
type ClientSocket {
  socketType SocketType
  *baseSocket
}

func NewClientSocket() Socket {
  return &ClientSocket{
    Type: ClientSocketType,
    baseSocket: &newBaseSocket(...),
  }
}

type SocketType string

```

The advantage of having an interface there is it will allow you to mock it easily in your tests if you need to. unlike a struct where you'd be forced to use the concrete implementation.